### PR TITLE
core: Add information on which condition failed to the job done message

### PR DESCRIPTION
**Pull request by daandemeyer**
When a job is skipped, it's useful to know exactly which condition failed so
let's add this information to the error message.

To avoid having to dynamically generate a format string, we special case the
formatting of condition failed messages.

### CODE REVIEW

File: src/core/job.c [hunk 1]
```diff
@@ -705,19 +705,45 @@ static void job_emit_done_message(Unit *u, uint32_t job_id, JobType t, JobResult
...
+                Condition *c;
                 const char *mid = job_done_mid(t, result);  /* mid may be NULL. log_unit_struct() will ignore it. */
-                const char *msg_fmt = strjoina("MESSAGE=", format);
 
-                DISABLE_WARNING_FORMAT_NONLITERAL;
-                log_unit_struct(u, job_done_messages[result].log_level,
-                                msg_fmt, ident,
+                LIST_FOREACH(conditions, c, u->conditions)
+                        if (c->result != CONDITION_SUCCEEDED)
+                                break;
```
Comment (poettering) [inline src/core/job.c]: maybe move this to some helper call:

```c
static Condition* unit_find_failed_condition(Unit *u) {
…
```

File: src/core/job.c [hunk 2]
```diff
@@ -705,19 +705,45 @@ static void job_emit_done_message(Unit *u, uint32_t job_id, JobType t, JobResult
...
+
+                if (t == JOB_START && result == JOB_DONE && !u->condition_result && c) {
```
Comment (poettering) [inline src/core/job.c]: maybe do the (possibly expensive) search for the failed condition only after checking the (cheap) JOB_START/JOB_DONE/condition_result, not before.

File: src/core/job.c [hunk 3]
```diff
@@ -705,19 +705,45 @@ static void job_emit_done_message(Unit *u, uint32_t job_id, JobType t, JobResult
...
+                        /* Special case units that were skipped because of a failed condition check so that
+                         * we can add more information to the message. */
+                        DISABLE_WARNING_FORMAT_NONLITERAL;
```
Comment (poettering) [inline src/core/job.c]: you should be able to drop these here, given you specify a literal format string here now instead of a variable

File: src/core/job.c [hunk 4]
```diff
@@ -705,19 +705,45 @@ static void job_emit_done_message(Unit *u, uint32_t job_id, JobType t, JobResult
...
+                                c->trigger ? "|" : "",
+                                c->negate ? "!" : "",
+                                c->parameter,
                                 "JOB_ID=%" PRIu32, job_id,
                                 "JOB_TYPE=%s", job_type_to_string(t),
                                 "JOB_RESULT=%s", job_result_to_string(result),
                                 LOG_UNIT_INVOCATION_ID(u),
                                 mid);
-                REENABLE_WARNING;
+                        REENABLE_WARNING;
```
Comment (poettering) [inline src/core/job.c]: as above

File: src/core/job.c (hunk 2 above)
Comment (daandemeyer) [reply src/core/job.c]: Moved the condition check and assignment to inside the if condition so we don't do it too early but also don't do it twice either. I'm not sure if we frown upon assignment in if conditions or if it's allowed though.

File: src/core/job.c [hunk 5]
```diff
@@ -705,19 +705,39 @@ static void job_emit_done_message(Unit *u, uint32_t job_id, JobType t, JobResult
...
                                * will actually print something to the console. */
-
+                Condition *c;
                 const char *mid = job_done_mid(t, result);  /* mid may be NULL. log_unit_struct() will ignore it. */
-                const char *msg_fmt = strjoina("MESSAGE=", format);
 
-                DISABLE_WARNING_FORMAT_NONLITERAL;
-                log_unit_struct(u, job_done_messages[result].log_level,
-                                msg_fmt, ident,
+                if (t == JOB_START && result == JOB_DONE && !u->condition_result && (c = unit_find_failed_condition(u))) {
```
Comment (yuwata) [inline src/core/job.c]: `{` is not necessary.

Maybe, it is better to move the condition `!u->condition_result` into `unit_find_failed_condition()`?

File: src/core/job.c [hunk 6]
```diff
@@ -705,19 +705,39 @@ static void job_emit_done_message(Unit *u, uint32_t job_id, JobType t, JobResult
...
                                * will actually print something to the console. */
-
+                Condition *c;
                 const char *mid = job_done_mid(t, result);  /* mid may be NULL. log_unit_struct() will ignore it. */
-                const char *msg_fmt = strjoina("MESSAGE=", format);
 
-                DISABLE_WARNING_FORMAT_NONLITERAL;
-                log_unit_struct(u, job_done_messages[result].log_level,
-                                msg_fmt, ident,
+                if (t == JOB_START && result == JOB_DONE && (c = unit_find_failed_condition(u)))
```
Comment (poettering) [inline src/core/job.c]: yeah, so I do frown about cramming too much stuff like this into one line. I mean, its not an iron rule, but doing multiple tests *and* an assignment is a bit much for me.

why not do:

```
Condition *c;

/* Figure out which condition failed in case the job was skipped due to a condition */
c = t == JOB_START && result == JOB_DONE && !u->condition_result ? 
                unit_find_failed_condition(u) : NULL;

if (c) {
        …
```

File: src/core/unit.c [hunk 7]
```diff
@@ -5848,6 +5848,19 @@ int unit_thaw_vtable_common(Unit *u) {
...
 
+Condition *unit_find_failed_condition(Unit *u) {
+        Condition *c;
+
+        if (!u->condition_result)
+                return NULL;
+
+        LIST_FOREACH(conditions, c, u->conditions)
+                if (c->result != CONDITION_SUCCEEDED)
+                        break;
```
Comment (poettering) [inline src/core/unit.c]: hmm, thinking about this more, I think this loop needs more love. The thing is that we have "triggering" conditions and regular ones. I think for the regular ones the current logic of returning immediately is OK, but for the "triggering" ones we should only return if we find no other reason to fail, because we combine the triggering conditions via a logic OR, instead of an AND. i.e. something like this should be more correct:

```c
        Condition *c, *failed_trigger = NULL;
        bool has_suceeded_trigger = false;
        LIST_FOREACH(conditions, c, u->conditions) {
                if (c->triggering) {
                        if (c->result == CONDITION_SUCCEEDED)
                                 has_suceeded_trigger = true;
                        else if (!failed_trigger)
                                 failed_trigger = c;
                } else (c->result != CONDITION_SUCCEEDED)
                        return c;
         }
        return failed_trigger && !has_suceeded_trigger ? failed_trigger : NULL;
```

Comment (daandemeyer) [reply src/core/unit.c]: Ahh, I had no idea about triggering conditions doing OR instead of AND. That makes sense, let me revise this

Comment (daandemeyer) [reply src/core/unit.c]: Hmm, thinking further, in the case of failed triggers, we might want to log all of the triggered conditions since logging only one of them won't really make sense.

Comment (poettering) [reply src/core/unit.c]: up to you. I think an "iterative" process for this is actually OK. where one condition fails and you fix it, and you then focus on the next and so on.

Comment (poettering) [reply src/core/unit.c]: i mean, such an "iterative" process is pretty simple to implement *and* to grok for users. So I think win/win for simplicity here

Comment (daandemeyer) [reply src/core/unit.c]: So I figure that if all triggers fail, we don't really need to log the triggers themselves since it's pretty trivial to just open the unit file and find them.

I did add a custom message for when all the triggers fail so we don't draw a user's attention to a single specific failed trigger condition but underline the fact that it's trigger conditions that failed and that not a single one of them succeeded.

File: src/core/job.c [hunk 8]
```diff
@@ -705,19 +705,53 @@ static void job_emit_done_message(Unit *u, uint32_t job_id, JobType t, JobResult
...
+                                        ident,
+                                        condition_type_to_string(c->type),
+                                        c->trigger ? "|" : "",
+                                        c->negate ? "!" : "",
+                                        c->parameter,
+                                        "JOB_ID=%" PRIu32, job_id,
+                                        "JOB_TYPE=%s", job_type_to_string(t),
+                                        "JOB_RESULT=%s", job_result_to_string(result),
+                                        LOG_UNIT_INVOCATION_ID(u),
+                                        mid);
```
Comment (poettering) [inline src/core/job.c:739]: uff, you have nested two if/else expressions, and did not use {} for that. I find that hard to follow. I mean, yes, we have the rule of not using {} for single-line if blocks, but I'd argue that this here is not single-line anymore, even if the compiler allows to drop the {} here... hence, please wrap the inner if/else in a {} to remove any confusion which else belongs to what.

File: src/core/job.c [hunk 9]
```diff
@@ -705,19 +705,53 @@ static void job_emit_done_message(Unit *u, uint32_t job_id, JobType t, JobResult
...
+                        else
+                                log_unit_struct(
+                                        u,
+                                        job_done_messages[result].log_level,
+                                        "MESSAGE=%s was skipped because of a failed condition check (%s=%s%s%s).",
+                                        ident,
+                                        condition_type_to_string(c->type),
+                                        c->trigger ? "|" : "",
```
Comment (poettering) [inline src/core/job.c]: this line can be dropped now

File: src/core/job.c [hunk 10]
```diff
@@ -705,19 +705,52 @@ static void job_emit_done_message(Unit *u, uint32_t job_id, JobType t, JobResult
...
-                DISABLE_WARNING_FORMAT_NONLITERAL;
-                log_unit_struct(u, job_done_messages[result].log_level,
-                                msg_fmt, ident,
-                                "JOB_ID=%" PRIu32, job_id,
-                                "JOB_TYPE=%s", job_type_to_string(t),
-                                "JOB_RESULT=%s", job_result_to_string(result),
-                                LOG_UNIT_INVOCATION_ID(u),
-                                mid);
-                REENABLE_WARNING;
+                c = t == JOB_START && result == JOB_DONE && !u->condition_result ?
```
Comment (yuwata) [inline src/core/job.c]: `!u->condition_result` can be dropped.

File: src/core/unit.c [hunk 11]
```diff
@@ -5848,6 +5848,25 @@ int unit_thaw_vtable_common(Unit *u) {
         return unit_cgroup_freezer_action(u, FREEZER_THAW);
 }
 
+Condition *unit_find_failed_condition(Unit *u) {
+        Condition *c, *failed_trigger = NULL;
+        bool has_suceeded_trigger = false;
```
Comment (yuwata) [inline src/core/unit.c]: suceeded -> succeeded

**Comment by poettering** [comment]
packit CI failure seems unrelated.
